cpp/src/arrow/dataset/filter_test.cc

```diff
@@ -0,0 +1,246 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/dataset/filter.h"
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "arrow/compute/api.h"
+#include "arrow/dataset/test_util.h"
+#include "arrow/record_batch.h"
+#include "arrow/status.h"
+#include "arrow/testing/gtest_util.h"
+
+namespace arrow {
+namespace dataset {
+
+using string_literals::operator""_;
+using internal::checked_cast;
+using internal::checked_pointer_cast;
+
+class ExpressionsTest : public ::testing::Test {
+ public:
+  void AssertSimplifiesTo(const Expression& expr, const Expression& given,
+                          const Expression& expected) {
+    auto simplified = expr.Assume(given);
+    ASSERT_OK(simplified.status());
+    if (!simplified.ValueOrDie()->Equals(expected)) {
+      FAIL() << "  simplification of: " << expr.ToString() << std::endl
+             << "              given: " << given.ToString() << std::endl
+             << "           expected: " << expected.ToString() << std::endl
+             << "                was: " << simplified.ValueOrDie()->ToString();
+    }
+  }
+
+  template <typename NnaryExpression, typename... T>
+  void AssertOperandsAre(const NnaryExpression& expr, ExpressionType::type type,
+                         T... expected_operands) {
+    ASSERT_EQ(expr.type(), type);
+    ASSERT_EQ(expr.operands().size(), sizeof...(T));
+    std::shared_ptr<Expression> expected_operand_ptrs[] = {expected_operands.Copy()...};
+
+    for (size_t i = 0; i < sizeof...(T); ++i) {
+      ASSERT_TRUE(expr.operands()[i]->Equals(expected_operand_ptrs[i]));
+    }
+  }
+
+  std::shared_ptr<ScalarExpression> always = ScalarExpression::Make(true);
+  std::shared_ptr<ScalarExpression> never = ScalarExpression::Make(false);
+};
+
+TEST_F(ExpressionsTest, Equality) {
+  ASSERT_TRUE("a"_.Equals("a"_));
+  ASSERT_FALSE("a"_.Equals("b"_));
+
+  ASSERT_TRUE(("b"_ == 3).Equals("b"_ == 3));
+  ASSERT_FALSE(("b"_ == 3).Equals("b"_ < 3));
+  ASSERT_FALSE(("b"_ == 3).Equals("b"_));
+
+  // ordering matters
+  ASSERT_FALSE(("b"_ > 2 and "b"_ < 3).Equals("b"_ < 3 and "b"_ > 2));
```

Comment:
Is there a particular reason for using "and" rather than "&&"?

Reply:
Well, I learnt something. I didn't know that these "alternate operator spellings" existed in C++.